validator_client/src/keystore/crypto.rs

```diff
@@ -0,0 +1,170 @@
+use crate::keystore::checksum::{Checksum, ChecksumModule};
+use crate::keystore::cipher::{Cipher, CipherModule};
+use crate::keystore::kdf::{Kdf, KdfModule};
+use serde::{Deserialize, Serialize};
+
+/// Crypto module for keystore.
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct Crypto {
+    pub kdf: KdfModule,
+    pub checksum: ChecksumModule,
+    pub cipher: CipherModule,
+}
+
+impl Crypto {
+    /// Generate crypto module for `Keystore` given the password,
+    /// secret to encrypt, kdf params and cipher params.
+    pub fn encrypt(password: String, secret: &[u8], kdf: Kdf, cipher: Cipher) -> Self {
+        // Generate derived key
+        let derived_key = match &kdf {
+            Kdf::Pbkdf2(pbkdf2) => pbkdf2.derive_key(&password),
+            Kdf::Scrypt(scrypt) => scrypt.derive_key(&password),
+        };
+        // Encrypt secret
+        let cipher_message = match &cipher {
+            Cipher::Aes128Ctr(cipher) => cipher.encrypt(&derived_key[0..16], secret),
+        };
+        // Generate checksum
+        let mut pre_image: Vec<u8> = derived_key[16..32].to_owned();
```

Comment:
I think we need a vector here since we don't know the size of the `secret` in advance. 
We can do this if we assume that the `secret` will always be a secret key of fixed length.

Reply:
Oh yeah, you're right sorry! I thought it was producing a HMAC but it's encrypting :)